File: queue/reveal_cards_in_increasing_order.cpp
class Solution {
public:
    vector<int> deckRevealedIncreasing(vector<int>& deck) {
        int n = deck.size();
        if (n == 1) return deck;

        sort(deck.begin(), deck.end());

        vector<int> result(n);
        queue<int> indexes;

        for (int i = 0; i < n; ++i) indexes.push(i);

        for (const int& card : deck) {
            result[indexes.front()] = card;
            indexes.pop();

            if (!indexes.empty()) {
                indexes.push(indexes.front());
                indexes.pop();
            }
        }

        return result;
    }
};

File: queue/reveal_cards_in_increasing_order.cs
public class Solution {
    public int[] DeckRevealedIncreasing(int[] deck) {
        if (deck.Length == 1) return deck;

        Array.Sort(deck);

        int[] result = new int[deck.Length];
        Queue<int> indexes = [];

        for (int i = 0; i < deck.Length; ++i) indexes.Enqueue(i);

        foreach (int card in deck) {
            result[indexes.Dequeue()] = card;
            if (indexes.Any()) indexes.Enqueue(indexes.Dequeue());
        }

        return result;
    }
}
